app/PaymentDrivers/Blockonomics/Blockonomics.php

```diff
@@ -144,41 +146,68 @@ public function paymentResponse(PaymentResponseRequest $request)
         ]);
 
         try {
-            $data = [];
             $fiat_amount = round(($request->btc_price * $request->btc_amount / 100000000), 2);
-            $data['amount'] = $fiat_amount;
-            $data['payment_method_id'] = $request->payment_method_id;
-            $data['payment_type'] = PaymentType::CRYPTO;
-            $data['gateway_type_id'] = GatewayType::CRYPTO;
+
+            $data = [
+                'amount' => $fiat_amount,
+                'payment_method_id' => $request->payment_method_id,
+                'payment_type' => PaymentType::CRYPTO,
+                'gateway_type_id' => GatewayType::CRYPTO,
+            ];
 
             // Append a random value to the transaction reference for test payments
             // to prevent duplicate entries in the database.
-            // This ensures the payment hashed_id remains unique.
             $testTxid = $this->test_txid;
             $data['transaction_reference'] = ($request->txid === $testTxid)
                 ? $request->txid . bin2hex(random_bytes(16))
                 : $request->txid;
 
-            $statusId = Payment::STATUS_PENDING;
-
-            switch ($request->status) {
-                case 0:
-                    $statusId = Payment::STATUS_PENDING;
-                    break;
-                case 1:
-                    $statusId = Payment::STATUS_PENDING;
-                    break;
-                case 2:
-                    $statusId = Payment::STATUS_COMPLETED;
-                    break;
-                default:
-                    $statusId = Payment::STATUS_PENDING;
+            // Determine payment status
+            $statusId = match($request->status) {
+                2 => Payment::STATUS_COMPLETED,
+                default => Payment::STATUS_PENDING
+            };
+
+            // Get payment hash and invoice before creating payment
+            $payment_hash = PaymentHash::where('hash', $request->payment_hash)->firstOrFail();
+            $invoice = $payment_hash->fee_invoice ?? $payment_hash->paymentable;
+
+            if (!$invoice) {
```

Comment:
every payment sequence _must_ have a payment hash, you should never reach a situation where you have to check for a payment has